src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -223,64 +231,193 @@ private UUID getParticipantEntitySetId( UUID studyId ) {
+         * Ideally we should only record UsageStats for apps that the user actually used
+         * Method getTotalTimeVisible() available in API level 29 would be the ideal filter since it gives the
+         * total time a package's activity was visible in the UI. However, our android app currently supports minimum API
+         * level 21. so the next best option would be getTotalTimeInForeground() which gives the total time spent in the
+         * foreground, measured in milliseconds
+         *
+         * https://developer.android.com/reference/android/app/usage/UsageStats
+         */
+        List<SetMultimap<UUID, Object>> userAppsData = data
+                .stream()
+                .filter( entry -> entry.containsKey( durationPropertyTypeId ) &&
+                        entry.containsKey( startDateTimePropertyId ) &&
+                        entry.containsKey( appNamePropertyTypeId ) &&
+                        Long.parseLong( entry.get( durationPropertyTypeId ).iterator().next().toString() ) > 0 )
+                .collect( Collectors.toList() );
+
+        for ( int i = 0; i < userAppsData.size(); ++i ) {
+
+            Set<DataEdgeKey> dataEdgeKeys = new HashSet<>();
+            String appPackageName = userAppsData.get( i ).get( appPackageNamePropertyTypeId ).iterator().next()
+                    .toString();
+            String dateLogged = parseDateTime( userAppsData.get( i ).get( startDateTimePropertyId ).iterator().next()
+                    .toString() );
+
+            // create entity in chronicle_user_apps
+            Map<UUID, Set<Object>> userAppEntityData = new HashMap<>();
+            userAppEntityData.put( appPackageNamePropertyTypeId, Sets.newHashSet( appPackageName ) );
+            userAppEntityData.put( appNamePropertyTypeId,
+                    Sets.newHashSet( userAppsData.get( i ).get( appNamePropertyTypeId ).iterator().next()
+                            .toString() ) );
+
+            UUID userAppEntityKeyId = reserveUserAppEntityKeyId( userAppEntityData, dataIntegrationApi );
+            dataApi.updateEntitiesInEntitySet( userAppsEntitySetId,
+                    ImmutableMap.of( userAppEntityKeyId, userAppEntityData ),
+                    UpdateType.Merge );
+
+            // association: chronicle_user_apps => chronicle_recorded_by => chronicle_device
+            Map<UUID, Set<Object>> recordedByEntityData = new HashMap<>();
+            recordedByEntityData.put( dateLoggedPropertyTypeId, ImmutableSet.of( dateLogged ) );
+            recordedByEntityData.put( stringIdPropertyTypeId, ImmutableSet.of( deviceId ) );
+
+            UUID recordedByEntityKeyId = reserveRecordedByEntityKeyId( recordedByEntityData,
+                    appPackageName,
+                    dataIntegrationApi );
+            dataApi.updateEntitiesInEntitySet( recordedByEntitySetId,
+                    ImmutableMap.of( recordedByEntityKeyId, recordedByEntityData ),
+                    UpdateType.Merge );
```

Comment:
Why `UpdateType.Merge`?

Reply:
I think any of the `UpdateType`s  are interchangeable in this situation.  The `recordedEntityKeyId` will be unique for a given device + app + date combination. The call to `dataApi.updateEntitiesInEntitySet` will create the entity if it does not exist, or update the property values if the entity exists. In reality though, for a given entity key id, the data is never updated since it's always the same.